### PR TITLE
rollback server-url sync in mcm settings handler

### PR DESCRIPTION
**Pull request by ibrokethecloud**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**

PR rolls back the server-url sync setting handler.

**Solution:**

**Related Issue:**
https://github.com/harvester/harvester/issues/4519

**Test plan:**

### CODE REVIEW

**Comment by ibrokethecloud** [comment]
The change will ensure that the rancher setting `server-url` is not set to ingress-expose external ip.

### Validation for new installs:
* install harvester 
* once cluster has bootstrapped download kubeconfig and verify `server-url` setting as follows
```
# This should be empty
kubectl get setting.management server-url
```

### Upgrade without custom ssl-certs
* install harvester v1.1.2
* once installation is completed trigger update to v1.2.1
* upgrade should complete successfully

### Upgrade with custom ssl-certs
* install harvester v1.1.2
* define setting for `ssl-certs`. ensure a valid cert chain including root-ca is present
* trigger upgrade to v1.2.1
* upgrade should complete successfully
* Issue mentioned in https://github.com/harvester/harvester/issues/4519 should not be encountered

### Add node to cluster with custom certs
* install harvester v1.2.0
* define setting for `ssl-certs`. ensure a valid cert chain including root-ca is present
* add another node to this cluster
* node should be added successfully and user not run into issue https://github.com/harvester/harvester/issues/4511

**Comment by ibrokethecloud** [comment]
2nd commit includes an additional mcmsetting handler to sync `internal-server-url` and `internal-cacerts` to the `fleet-controller` configmap in the `cattle-fleet-system` namespace. 

The packaged rancher will only be managing the local cluster only. So this should ensure that changes to these settings are always reconciled to the configmap.

Addresses: https://github.com/harvester/harvester/issues/4565

File: pkg/controller/master/mcmsettings/fleetcontroller.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	"github.com/rancher/wrangler/pkg/relatedresource"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	defaultFleetControllerConfigMapName      = "fleet-controller"
+	defaultFleetControllerConfigMapNamespace = "cattle-fleet-system"
```
Comment (w13915984028) [inline pkg/controller/master/mcmsettings/fleetcontroller.go]: a better place for namespace definition is the `pkg/util/constants.go`

File: pkg/controller/master/mcmsettings/fleetcontroller.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	var patchedConfigMap *corev1.ConfigMap
+	if currentVal != encodedCA {
+		patchedConfigMap, err = patchConfigMap(fleetControllerConfig, apiServerCAKey, encodedCA)
+		if err != nil {
+			return setting, err
+		}
+
+		_, err = h.configMapClient.Update(patchedConfigMap)
```
Comment (w13915984028) [inline pkg/controller/master/mcmsettings/fleetcontroller.go:72]: Please double check:

I did some manual tests again `fleet-controller` configmap, it seems that the change to this configmap will not cause helm chart `fleet-controller` to be re-deployed. 

If we delete the deployment of `fleet-controller` and update any of rancher below settings, the `fleet-controller` will be re-deployed, the new configmap is used then.
```
	if setting.Name != settings.ServerURL.Name &&
		setting.Name != settings.CACerts.Name &&
		setting.Name != settings.SystemDefaultRegistry.Name {
		return setting, nil
	}
```

The related code is:

https://github.com/rancher/rancher/blob/release/v2.7.6/pkg/controllers/dashboard/fleetcharts/controller.go#L60

**Review by w13915984028 — COMMENTED** [review]
Please check, thanks.

File: pkg/controller/master/mcmsettings/fleetcontroller.go (hunk 2 above)
Comment (ibrokethecloud) [reply pkg/controller/master/mcmsettings/fleetcontroller.go:72]: The aim of the change is to not redeploy the fleet-controller. Rancher manages the fleet controller update as part of the upgrade process. Fleet controller then manages the upgrade of downstream clusters fleet-agent. This change ensures that fleet-agent is redeployed by the fleet-controller.

Currently the fleet-controller never reconciles the cluster objects because it cannot find the `apiServerURL` defined in the config map.

The change ensures this is set, which allows the fleet controller to reconcile the imported cluster, and subsequently redeploy the fleet-agent.

https://github.com/rancher/fleet/blob/v0.7.0/pkg/controllers/cluster/import.go#L235

File: pkg/controller/master/mcmsettings/fleetcontroller.go (hunk 1 above)
Comment (ibrokethecloud) [reply pkg/controller/master/mcmsettings/fleetcontroller.go]: good point.. i will move it there.

File: pkg/controller/master/mcmsettings/fleetcontroller.go (hunk 2 above)
Comment (w13915984028) [reply pkg/controller/master/mcmsettings/fleetcontroller.go:72]: Sounds reasonable.

Should the upstream Rancher add this enhancement formally later ? From the view of Harvester, this workaround has potential conflicts with Rancher & Fleet-controller in case their logic changes in the future.

File: pkg/controller/master/mcmsettings/fleetcontroller.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+func (h *mcmSettingsHandler) watchInternalEndpointSettings(_ string, setting *mgmtv3.Setting) (*mgmtv3.Setting, error) {
+	if setting == nil || setting.DeletionTimestamp != nil || (setting.Name != util.RancherInternalCASetting && setting.Name != util.RancherInternalServerURLSetting) {
+		return setting, nil
+	}
+
+	logrus.Debugf("reconcilling fleet-controller config for setting %s", setting.Name)
+
+	fleetControllerConfig, err := h.configMapCache.Get(util.DefaultFleetControllerConfigMapNamespace, util.DefaultFleetControllerConfigMapName)
+	if err != nil {
```
Comment (w13915984028) [inline pkg/controller/master/mcmsettings/fleetcontroller.go:42]: Please consider the posssible future case:

When the upstream Rancher does not use the default namespace/name. Here we may use h.configMap.Get() and check if the error is IsNotExisting, then log a warning and return nil.

**Comment by w13915984028** [comment]
Another question is about the second commit:

We add processing for below Rancher settings, in Harvester's context, how to trigger the change of them and then validate ? thanks.

```
	RancherInternalCASetting                 = "internal-cacerts"
	RancherInternalServerURLSetting          = "internal-server-url"
```

File: pkg/controller/master/mcmsettings/fleetcontroller.go (hunk 2 above)
Comment (ibrokethecloud) [reply pkg/controller/master/mcmsettings/fleetcontroller.go:72]: we do have a related issue for this: https://github.com/rancher/rancher/issues/41848

File: pkg/controller/master/mcmsettings/fleetcontroller.go (hunk 3 above)
Comment (ibrokethecloud) [reply pkg/controller/master/mcmsettings/fleetcontroller.go:42]: the current fleet-controller design needs the config map to exist. If this changes we can accordingly update or remove this handler if this is not needed.

**Comment by ibrokethecloud** [comment]
rancher generates and sets those settings. I am not sure if there is an easy way to trigger the update. It seems to set it to the cluster IP for rancher svc.

**Comment by w13915984028** [comment]
I suspect the `internal-server-url` will not change in the life cycle of a cluster. To be confirmed.

**Comment by ibrokethecloud** [comment]
this should address the issue: https://github.com/rancher/rancher/pull/42649

we can rollback this and rancherd changes once this is available in a release

**Review by bk201 — APPROVED** [review]
lgtm!

**Review by w13915984028 — APPROVED** [review]
LGTM, thanks.

**Review by starbops — APPROVED** [review]
I've tested the rollback with three cases. Two for the node-joining and one for the upgrading. And from the scope of the issue this PR intended to solve, it LGTM.

TLDR

1. Joining node with FQDN will fail at `rancher-system-agent`, need to apply a workaround
2. Joining node with VIP address will fail at `rancherd`, need to apply a workaround, too
3. Upgrading looks good

Details are provided below.

---

Premises: the running cluster has already configure with a set of self-signed root CA and certificate (without IP SANs extension)

**Case 1: Joining node with FQDN**

Adding nodes with an FQDN, in this case, `harvester.192.168.48.240.sslip.io`, as the management address into a cluster with a custom SSL certificate configured failed at starting up `rancher-system-agent` due to the "unknown authority" issue
    
From `rancherd`'s log, it passed the bootstrapping stage and was waiting for the kubelet to be online:
    
```bash 
$ sudo journalctl -u rancherd.service --no-pager
Oct 02 02:34:16 node-0 systemd[1]: Starting Rancher Bootstrap...
Oct 02 02:34:16 node-0 rancherd[2022]: time="2023-10-02T02:34:16Z" level=info msg="Loading config file [/usr/share/rancher/rancherd/config.yaml.d/50-defaults.yaml]"
Oct 02 02:34:16 node-0 rancherd[2022]: time="2023-10-02T02:34:16Z" level=info msg="Loading config file [/usr/share/rancher/rancherd/config.yaml.d/91-harvester-bootstrap-repo.yaml]"
Oct 02 02:34:16 node-0 rancherd[2022]: time="2023-10-02T02:34:16Z" level=info msg="Loading config file [/etc/rancher/rancherd/config.yaml]"
Oct 02 02:34:16 node-0 rancherd[2022]: time="2023-10-02T02:34:16Z" level=info msg="Bootstrapping Rancher (v2.7.5/v1.25.9+rke2r1)"
Oct 02 02:34:17 node-0 rancherd[2022]: time="2023-10-02T02:34:17Z" level=info msg="Writing plan file to /var/lib/rancher/rancherd/plan/plan.json"
Oct 02 02:34:17 node-0 rancherd[2022]: time="2023-10-02T02:34:17Z" level=info msg="Applying plan with checksum "
Oct 02 02:34:17 node-0 rancherd[2022]: time="2023-10-02T02:34:17Z" level=info msg="No image provided, creating empty working directory /var/lib/rancher/rancherd/plan/work/20231002-023417-applied.plan/_0"
Oct 02 02:34:17 node-0 rancherd[2022]: time="2023-10-02T02:34:17Z" level=info msg="Running command: /usr/bin/env [sh /var/lib/rancher/rancherd/install.sh]"
Oct 02 02:34:17 node-0 rancherd[2022]: time="2023-10-02T02:34:17Z" level=info msg="[stdout]: [INFO]  Using default agent configuration directory /etc/rancher/agent"
Oct 02 02:34:17 node-0 rancherd[2022]: time="2023-10-02T02:34:17Z" level=info msg="[stdout]: [INFO]  Using default agent var directory /var/lib/rancher/agent"
Oct 02 02:34:17 node-0 rancherd[2022]: time="2023-10-02T02:34:17Z" level=info msg="[stderr]: [WARN]  /usr/local is read-only or a mount point; installing to /opt/rancher-system-agent"
Oct 02 02:34:17 node-0 rancherd[2022]: time="2023-10-02T02:34:17Z" level=info msg="[stdout]: [INFO]  Determined CA is necessary to connect to Rancher"
Oct 02 02:34:18 node-0 rancherd[2022]: time="2023-10-02T02:34:18Z" level=info msg="[stdout]: [INFO]  Successfully downloaded CA certificate"
Oct 02 02:34:18 node-0 rancherd[2022]: time="2023-10-02T02:34:18Z" level=info msg="[stdout]: [INFO]  Value from https://harvester.192.168.48.240.sslip.io:443/cacerts is an x509 certificate"
Oct 02 02:34:18 node-0 rancherd[2022]: time="2023-10-02T02:34:18Z" level=info msg="[stdout]: [INFO]  Successfully tested Rancher connection"
Oct 02 02:34:18 node-0 rancherd[2022]: time="2023-10-02T02:34:18Z" level=info msg="[stdout]: [INFO]  Downloading rancher-system-agent binary from https://harvester.192.168.48.240.sslip.io:443/assets/rancher-system-agent-amd64"
Oct 02 02:34:18 node-0 rancherd[2022]: time="2023-10-02T02:34:18Z" level=info msg="[stdout]: [INFO]  Successfully downloaded the rancher-system-agent binary."
Oct 02 02:34:18 node-0 rancherd[2022]: time="2023-10-02T02:34:18Z" level=info msg="[stdout]: [INFO]  Downloading rancher-system-agent-uninstall.sh script from https://harvester.192.168.48.240.sslip.io:443/assets/system-agent-uninstall.sh"
Oct 02 02:34:18 node-0 rancherd[2022]: time="2023-10-02T02:34:18Z" level=info msg="[stdout]: [INFO]  Successfully downloaded the rancher-system-agent-uninstall.sh script."
Oct 02 02:34:18 node-0 rancherd[2022]: time="2023-10-02T02:34:18Z" level=info msg="[stdout]: [INFO]  Generating Cattle ID"
Oct 02 02:34:19 node-0 rancherd[2022]: time="2023-10-02T02:34:19Z" level=info msg="[stdout]: [INFO]  Successfully downloaded Rancher connection information"
Oct 02 02:34:19 node-0 rancherd[2022]: time="2023-10-02T02:34:19Z" level=info msg="[stdout]: [INFO]  systemd: Creating service file"
Oct 02 02:34:19 node-0 rancherd[2022]: time="2023-10-02T02:34:19Z" level=info msg="[stdout]: [INFO]  Creating environment file /etc/systemd/system/rancher-system-agent.env"
Oct 02 02:34:19 node-0 rancherd[2022]: time="2023-10-02T02:34:19Z" level=info msg="[stdout]: [INFO]  Enabling rancher-system-agent.service"
Oct 02 02:34:19 node-0 rancherd[2022]: time="2023-10-02T02:34:19Z" level=info msg="[stderr]: Created symlink /etc/systemd/system/multi-user.target.wants/rancher-system-agent.service → /etc/systemd/system/rancher-system-agent.service."
Oct 02 02:34:19 node-0 rancherd[2022]: time="2023-10-02T02:34:19Z" level=info msg="[stdout]: [INFO]  Starting/restarting rancher-system-agent.service"
Oct 02 02:34:19 node-0 rancherd[2022]: time="2023-10-02T02:34:19Z" level=info msg="No image provided, creating empty working directory /var/lib/rancher/rancherd/plan/work/20231002-023417-applied.plan/_1"
Oct 02 02:34:19 node-0 rancherd[2022]: time="2023-10-02T02:34:19Z" level=info msg="Running command: /usr/bin/rancherd [probe]"
Oct 02 02:34:19 node-0 rancherd[2022]: time="2023-10-02T02:34:19Z" level=info msg="[stderr]: time=\"2023-10-02T02:34:19Z\" level=info msg=\"Running probes defined in /var/lib/rancher/rancherd/plan/plan.json\""
Oct 02 02:34:20 node-0 rancherd[2022]: time="2023-10-02T02:34:20Z" level=info msg="[stderr]: time=\"2023-10-02T02:34:20Z\" level=info msg=\"Probe [kubelet] is unhealthy\""
```
    
However, the `rancher-system-agent` failed with the following error:
    
```bash
$ journalctl -u rancher-system-agent.service
Oct 02 02:54:09 node-0 systemd[1]: Started Rancher System Agent.
Oct 02 02:54:09 node-0 rancher-system-agent[6641]: time="2023-10-02T02:54:09Z" level=info msg="Rancher System Agent version v0.3.3 (9e827a5) is starting"
Oct 02 02:54:09 node-0 rancher-system-agent[6641]: time="2023-10-02T02:54:09Z" level=info msg="Using directory /var/lib/rancher/agent/work for work"
Oct 02 02:54:09 node-0 rancher-system-agent[6641]: time="2023-10-02T02:54:09Z" level=info msg="Starting remote watch of plans"
Oct 02 02:54:09 node-0 rancher-system-agent[6641]: time="2023-10-02T02:54:09Z" level=info msg="Initial connection to Kubernetes cluster failed with error Get \"https://harvester.192.168.48.240.sslip.io/version\": x509: certificate signed by unknown authority, removing CA data and trying again"
Oct 02 02:54:09 node-0 rancher-system-agent[6641]: time="2023-10-02T02:54:09Z" level=fatal msg="error while connecting to Kubernetes cluster with nullified CA data: Get \"https://harvester.192.168.48.240.sslip.io/version\": x509: certificate signed by unknown authority"
Oct 02 02:54:09 node-0 systemd[1]: rancher-system-agent.service: Main process exited, code=exited, status=1/FAILURE
Oct 02 02:54:09 node-0 systemd[1]: rancher-system-agent.service: Failed with result 'exit-code'.
```
    
The workaround is simple: add the CA certificate to the affected nodes:
    
```bash
sudo cp ca.crt /etc/pki/trust/anchors/
sudo update-ca-certificates
```

After that, you'll see the nodes joined into the cluster successfully.

**Case 2: Joining node with VIP address**

Adding nodes with a VIP address as the management address into a cluster with a custom SSL certificate configured will fail at starting `rancherd` due to the "no IP SANs" issue

```bash
$ journalctl -u rancherd.service
Oct 02 03:28:38 node-0 rancherd[2020]: time="2023-10-02T03:28:38Z" level=info msg="Loading config file [/usr/share/rancher/rancherd/config.yaml.d/50-defaults.yaml]"
Oct 02 03:28:38 node-0 rancherd[2020]: time="2023-10-02T03:28:38Z" level=info msg="Loading config file [/usr/share/rancher/rancherd/config.yaml.d/91-harvester-bootstrap-repo.yaml]"
Oct 02 03:28:38 node-0 rancherd[2020]: time="2023-10-02T03:28:38Z" level=info msg="Loading config file [/etc/rancher/rancherd/config.yaml]"
Oct 02 03:28:38 node-0 rancherd[2020]: time="2023-10-02T03:28:38Z" level=info msg="Bootstrapping Rancher (v2.7.5/v1.25.9+rke2r1)"
Oct 02 03:28:38 node-0 rancherd[2020]: time="2023-10-02T03:28:38Z" level=info msg="failed to bootstrap system, will retry: generating plan: Get \"https://192.168.48.240:443/system-agent-install.sh\": x509: cannot validate certificate for 192.168.48.240 because it doesn't contain any IP SANs"
```

The workaround is to change the `ssl-certificate` Setting of the Harvester cluster with the certificate containing valid IP SANs. After changing the certificate, the burden fell to `rancher-system-agent`:

```bash
$ journalctl -u rancher-system-agent.service
Oct 02 03:42:44 node-0 systemd[1]: rancher-system-agent.service: Scheduled restart job, restart counter is at 10.
Oct 02 03:42:44 node-0 systemd[1]: Stopped Rancher System Agent.
Oct 02 03:42:44 node-0 systemd[1]: Started Rancher System Agent.
Oct 02 03:42:44 node-0 rancher-system-agent[3789]: time="2023-10-02T03:42:44Z" level=info msg="Rancher System Agent version v0.3.3 (9e827a5) is starting"
Oct 02 03:42:44 node-0 rancher-system-agent[3789]: time="2023-10-02T03:42:44Z" level=info msg="Using directory /var/lib/rancher/agent/work for work"
Oct 02 03:42:44 node-0 rancher-system-agent[3789]: time="2023-10-02T03:42:44Z" level=info msg="Starting remote watch of plans"
Oct 02 03:42:44 node-0 rancher-system-agent[3789]: time="2023-10-02T03:42:44Z" level=info msg="Initial connection to Kubernetes cluster failed with error Get \"https://192.168.48.240/version\": x509: certificate signed by unknown authority, removing CA data and trying again"
Oct 02 03:42:44 node-0 rancher-system-agent[3789]: time="2023-10-02T03:42:44Z" level=fatal msg="error while connecting to Kubernetes cluster with nullified CA data: Get \"https://192.168.48.240/version\": x509: certificate signed by unknown authority"
Oct 02 03:42:44 node-0 systemd[1]: rancher-system-agent.service: Main process exited, code=exited, status=1/FAILURE
Oct 02 03:42:44 node-0 systemd[1]: rancher-system-agent.service: Failed with result 'exit-code'.
```

So, the last step of the workaround is the same as the previous case: adding the CA certificate to the joining nodes:

```bash
sudo cp ca.crt /etc/pki/trust/anchors/
sudo update-ca-certificates
```

And everything will be fine.

**Case 3: Upgrading**

1. Prepare a v1.1.2 single-node cluster
2. Generate a self-signed root CA and certificate (no IP SAN configured)
3. Set up `ssl-certificates` Setting CR with the generated private key, certificate, and CA
4. Kick-off the upgrade with the above ISO image
5. Wait until the upgrade ends successfully
6. Check the `cattle-fleet-local-system/fleet-agent` Pod's log. No error messages were found.
7. Check the `cattle-fleet-system/fleet-controller` Pod's log. No error messages were found.
8. The `bundles` statuses are all good
9. The `server-url` Setting is still empty
10. The `cacerts` Setting still has no value but a default value (`.default`) equal to the CA in `ssl-certificates`. Note: it is what it was before the upgrade, so that it may be no big deal.
11. There's no `fleet-agent-bootstrap` Secret
12. Check the `cattle-fleet-local-system/fleet-agent` Secret. The `.data.kubeconfig` field contains the correct server URL (same as the content of the `internal-server-url` Setting) and `certificate-authority-data` (same as the content of `internal-cacerts` Setting).
13. Check the `cattle-fleet-system/fleet-controller` ConfigMap. The `apiServerCA` is the same as `internal-cacerts`. The `apiServerURL` is the same as `internal-server-url`.
14. Check the `fleet-local/local-kubeconfig` Secret. The `.data.value` field contains the correct server URL (same as the content of the `internal-server-url` Setting) and `certificate-authority-data` (same as the content of `internal-cacerts` Setting).

**Comment by bk201** [comment]
@ibrokethecloud Can you help squash the commits? One for the rollback and one for the new controller.
We'll need to backport this to v1.2 and cherry-pick them.

**Comment by bk201** [comment]
@mergifyio backport v1.2
